File: src/pageviews/separate_pageviews.pl
#!/usr/bin/env perl

use File::Basename;
use File::Path qw(make_path);
use Env qw(HOSTNAME);
use Pageviews;
#use IO::Handle;
use IO::File;

#print "$HOSTNAME\$ " . basename($0) . " @ARGV\n";

# check the command-line arguments
die "usage: $0 [OUTPUT_PATH [FILE1 [FILE2 [...]]]]" if (scalar @ARGV < 1);

$output_path = 'pageviews';
if (@ARGV) {
	$output_path = shift @ARGV;
}
#print "The pageviews files are located at $output_path.\n";

#%titles = ();
#if (@ARGV) {
#	open FILE, "<$ARGV[0]" or die "failed opening $ARGV[0]: $!";
#	while (<FILE>) {
#		@fields = split;
#		if ($#fields > 0) {
#			# there are more than one fields, so this must not be a list of titles.
#			break;
#		}
#		chomp;
#		$titles{$_} = 1;
#	}
#	close FILE;
#
#	if (%titles) {
#		shift;
#	}
#}

# modules
@open_files=();
%files=();
sub open_and_append {
	my $output_path = shift;
	my $title = shift;
	die "title is empty!" if (!defined $title or $title eq "");

	if (exists $files{$title}) {
		return $files{$title};
	} else {
		if (scalar @open_files >= 100) {
			$oldest_title = shift @open_files;
			delete $files{$oldest_title};
		}
		push @open_files, $title;
		my $filename = File::Spec->join($output_path, title_to_filename($title));
#		my $fh;
#		if (!-e "$filename") {
#			$fh = IO::File->new(">$filename") or die "could not open $filename: $!";
#			print $fh "date\tpageview\n";
#		} else {
#			$fh = IO::File->new(">>$filename") or die "could not append $filename: $!";
#		}
		my $fh = IO::File->new(">>$filename") or die "could not append $filename: $!";
		$files{$title}=$fh;
		return $fh;
	}
}

sub separate_file {
	my $infh = shift;
	while (<$infh>) {
		@fields = split;
		$title = $fields[1];
		$fh = open_and_append($output_path, $title);
		print $fh "$fields[0]\t$fields[2]\n";
	}
	
}

if (!-e ($output_path)) {
	make_path($output_path) or die "creating $output_path failed: $!";
}

if (@ARGV) {
	while (@ARGV) {
		$pattern = shift @ARGV;
		foreach $file (glob("$pattern")) {
			print "separating $file\n";
			open my $fh, "<$file" or die "failed opening $file: $!";
			separate_file($fh, $title);
			close $fh;
		}
	}
} else {
	separate_file(STDIN);
}

for $fh (values %files) {
	$fh->close;
}

File: src/pageviews/separate_pageviews.py
#!/usr/bin/env python

import sys
import urllib
import hashlib
import os
import os.path
#import datetime.date

# the first argument is the name of a text file that contains the title of Wikipedia pages in the list one each line
# the second argument is the file that contains pageviews; the format is "date" "page title" "page view"

titles = {}

if len(sys.argv) > 1:
	output_path = sys.argv[1]
else:
	output_path = 'pageviews'

if len(sys.argv) > 2:
	list_file = sys.argv[2]
	if not os.path.exists(list_file):
		print list_file, 'does not exist'
		sys.exit(1)
	else:
		for lines in open(list_file):
			titles[lines.strip()] = True

if len(sys.argv) > 3:
	date = sys.argv[3]
else:
	#date = datetime.date.today().isoformat()
	date = 'unknown_date'

if not os.path.isdir(output_path):
	try:
		os.makedirs(output_path)
	except:
		sys.exit(1)
assert os.path.isdir(output_path)

def title_to_filename(title):
	# convert the title into one usable as filename
	filename = urllib.quote(title, safe="%") # force / to be quoted and % not to be quoted
	if filename.startswith('.'):
		filename = "%2E" + filename[1:]
	if len(filename) > 128:
		digest = hashlib.sha224(filename).hexdigest()
		digest = urllib.quote(digest, safe="%")
		filename = filename[:128] + digest
	filename = os.path.join(output_path, filename + '_pageviews.txt')
	return filename

def open_and_append(filename):
	firstline = not os.path.exists(filename)
	filehandle = open(filename, 'a')
	if firstline:
		filehandle.write('date\tpageview\n')
	return filehandle

try:
	for line in sys.stdin:
		fields = line.split()
		title = fields[1]
		pageview = fields[2]
		if title in titles:
			del titles[title]
		filehandle = open_and_append(title_to_filename(title))
		filehandle.write(date + '\t' + pageview + '\n')
		filehandle.close()
except:
	pass
finally:
	for title in titles.keys():
		filehandle = open_and_append(title_to_filename(title))
		filehandle.write(date + '\t0\n')
		filehandle.close()
